jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java

```diff
@@ -298,6 +301,70 @@ public EndPoint getEndPoint()
         return _endPoint;
     }
 
+    /**
+     * <p>
+     * Override the {@link #getLocalName()} with a specific name.
+     * </p>
+     *
+     * <p>
+     * This allows the name returned by APIs like HttpServletRequest.getServerName() to be overridden with
+     * an arbitrary name when behind an intermediary and the incoming request has no valid uri authority.
+     * </p>
+     *
+     * @param overriddenLocalName the overridden local name, use null to set behavior back to default
+     */
+    public void setLocalNameOverride(String overriddenLocalName)
+    {
+        this.overriddenLocalName = overriddenLocalName;
+    }
+
+    /**
+     * <p>Return the Local Name of the connected channel.</p>
+     *
+     * <p>
+     * This is the name after the connector is bound and the connection is accepted.
+     * The value is typically the local address and optionally the local host name
+     * if the local address is unavailable.
+     * </p>
+     * <p>
+     * Value can be overridden by {@link #setLocalNameOverride(String)} for
+     * scenarios where Jetty is being an intermediary and the local name
+     * needs to be changed to satisfy public (pre-intermediary) HTTP behaviors
+     * such as absolute-URI creation (eg: Location response header).
+     * </p>
+     *
+     * @return the local name if overridden, or the local address, or
+     * the local host name.
+     */
+    public String getLocalName()
+    {
+        Connector connector = getConnector();
+
+        if (connector instanceof AbstractConnector)
+        {
+            String localName = ((AbstractConnector)connector).getLocalName();
+            if (localName != null)
+                return localName;
+        }
+
+        InetSocketAddress local = getLocalAddress();
+        if (local != null)
+            return local.getHostString();
+
+        try
+        {
+            String name = InetAddress.getLocalHost().getHostName();
+            if (StringUtil.ALL_INTERFACES.equals(name))
+                return null;
+            return name;
+        }
+        catch (UnknownHostException e)
+        {
+            LOG.ignore(e);
+        }
```

Comment:
Don't think this code should be there.
It's typically wrong if `/etc/hosts` is not setup properly, it does not work for Unix-Domain, leaks the internal host name or IP, etc.